# Employee Attrition Intelligence System

## Notebook 01: Data Audit & Cleaning

### Objective

The purpose of this notebook is to:

- Load the IBM HR Attrition dataset
- Understand dataset structure
- Assess data quality
- Identify missing values
- Check for duplicates
- Validate data types
- Detect potential outliers
- Prepare a clean dataset for analysis

### Expected Output

A cleaned and validated dataset ready for:
- SQL Analysis
- Exploratory Data Analysis (EDA)
- Feature Engineering
- Machine Learning
- Power BI Dashboarding

## 1. Import Libraries

In [3]:
import pandas as pd

print(pd.__version__)

3.0.3


## 2. Load Dataset

In [4]:
import pandas as pd

df = pd.read_csv("../data/raw/WA_Fn-UseC_-HR-Employee-Attrition.csv")

print("Dataset loaded successfully")

Dataset loaded successfully


## 3. Dataset Dimensions

Understanding the number of records and features helps establish
the scope of the analysis and verify that the dataset has been loaded correctly.

In [5]:
print("Rows:", df.shape[0])
print("Columns:", df.shape[1])

Rows: 1470
Columns: 35


## 4. Dataset Structure

Understanding the dataset structure helps identify:

- Available features
- Data types
- Potential data quality issues
- Columns requiring transformation

This step provides an overview of the dataset before performing detailed data quality checks.

In [6]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 1470 entries, 0 to 1469
Data columns (total 35 columns):
 #   Column                    Non-Null Count  Dtype
---  ------                    --------------  -----
 0   Age                       1470 non-null   int64
 1   Attrition                 1470 non-null   str  
 2   BusinessTravel            1470 non-null   str  
 3   DailyRate                 1470 non-null   int64
 4   Department                1470 non-null   str  
 5   DistanceFromHome          1470 non-null   int64
 6   Education                 1470 non-null   int64
 7   EducationField            1470 non-null   str  
 8   EmployeeCount             1470 non-null   int64
 9   EmployeeNumber            1470 non-null   int64
 10  EnvironmentSatisfaction   1470 non-null   int64
 11  Gender                    1470 non-null   str  
 12  HourlyRate                1470 non-null   int64
 13  JobInvolvement            1470 non-null   int64
 14  JobLevel                  1470 non-null   int64
 15

## 5. Missing Value Assessment

Missing values can impact analysis, feature engineering, and model performance.

This step verifies whether any columns contain missing values and quantifies their extent.

In [9]:
missing_values = df.isnull().sum()

missing_summary = pd.DataFrame({
    "Missing Values": missing_values,
    "Missing Percentage": (missing_values / len(df)) * 100
})

missing_columns = missing_summary[
    missing_summary["Missing Values"] > 0
]

print("Columns with missing values:", len(missing_columns))

missing_columns

Columns with missing values: 0


,Missing Values,Missing Percentage


### Findings

- No missing values were detected in the dataset.
- All 35 columns contain complete records.
- No imputation or missing value treatment is required.

## 6. Duplicate Record Assessment

Duplicate employee records can distort analysis results and model training.

This step verifies whether duplicate rows exist in the dataset.

In [10]:
duplicate_count = df.duplicated().sum()

print("Duplicate Records:", duplicate_count)

Duplicate Records: 0


### Findings

- No duplicate employee records were identified.
- The dataset contains 1,470 unique employee records.
- Duplicate removal is not required.

## 7. Initial Data Validation

This step reviews sample records to verify that columns contain reasonable values and to identify any potential data quality concerns.

In [11]:
df.head()

,Age,Attrition,BusinessTravel,DailyRate,Department,DistanceFromHome,Education,EducationField,EmployeeCount,EmployeeNumber,...,RelationshipSatisfaction,StandardHours,StockOptionLevel,TotalWorkingYears,TrainingTimesLastYear,WorkLifeBalance,YearsAtCompany,YearsInCurrentRole,YearsSinceLastPromotion,YearsWithCurrManager
0,41,Yes,Travel_Rarely,1102,Sales,1,2,Life Sciences,1,1,...,1,80,0,8,0,1,6,4,0,5
1,49,No,Travel_Frequently,279,Research & Development,8,1,Life Sciences,1,2,...,4,80,1,10,3,3,10,7,1,7
2,37,Yes,Travel_Rarely,1373,Research & Development,2,2,Other,1,4,...,2,80,0,7,3,3,0,0,0,0
3,33,No,Travel_Frequently,1392,Research & Development,3,4,Life Sciences,1,5,...,3,80,0,8,3,3,8,7,3,0
4,27,No,Travel_Rarely,591,Research & Development,2,1,Medical,1,7,...,4,80,1,6,3,3,2,2,2,2


## 8. Constant Value Assessment

Columns containing only a single value do not provide analytical or predictive value.

This step identifies features with no variability.

In [12]:
constant_columns = []

for col in df.columns:
    if df[col].nunique() == 1:
        constant_columns.append(col)

print("Constant Columns:")
print(constant_columns)

Constant Columns:
['EmployeeCount', 'Over18', 'StandardHours']


### Findings

Three constant-value columns were identified:

- EmployeeCount
- Over18
- StandardHours

These columns contain no variability across records and therefore do not contribute any predictive or analytical value.

They will be removed in the feature engineering phase to improve model efficiency and reduce noise.

## 9. Data Quality Summary

Based on initial audit:

- No missing values
- No duplicate records
- 3 constant-value columns identified
- Dataset is structurally clean and analysis-ready

Next step: Exploratory Data Analysis (EDA) to understand relationships between features and attrition.